File: 125.valid-palindrome.go
func realIsPalindrome(s string) bool {
	for i := 0; i < len(s)/2; i++ {
		if s[i] != s[len(s)-1-i] {
			return false
		}
	}

	return true
}

func isPalindrome(s string) bool {
	t := ""
	for i := 0; i < len(s); i++ {
		if 'A' <= s[i] && s[i] <= 'Z' {
			t += string(s[i] - 'A' + 'a')
		}
		if 'a' <= s[i] && s[i] <= 'z' || '0' <= s[i] && s[i] <= '9' {
			t += string(s[i])
		}
	}

	return realIsPalindrome(t)
}

File: 125.valid-palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int l = 0;
        int r = s.size()-1;

        while (l < r) {
            if (!check(s[l])) {
                l += 1;
                continue;
            }
            if (!check(s[r])) {
                r -= 1;
                continue;
            }
            if (tolower(s[l]) == tolower(s[r])) {
                l += 1;
                r -= 1;
            } else {
                return false;
            }
        }

        return true;
    }

    bool check(char c) {
        if ('a' <= c && c <= 'z') {
            return true;
        }

        if ('A' <= c && c <= 'Z') {
            return true;
        }

        if ('0' <= c && c <= '9') {
            return true;
        }

        return false;
    }
};
